articles/how-to-get-ssl-https-for-localhost/index.md

```diff
@@ -0,0 +1,200 @@
+```
+
+The information that needs to be written into the signed SSL certificate will be contained in this `localhost.ext` file. The file will appear as below:
+
+```
+authorityKeyIdentifier = keyid,issuer
+basicConstraints = CA:FALSE
+keyUsage = digitalSignature, nonRepudiation, keyEncipherment, dataEncipherment
+subjectAltName = @alt_names
+
+[alt_names]
+DNS.1 = localhost
+IP.1 = 127.0.0.1
+```
+
+Note that the certificate will work for localhost and also 127.0.0.1. One can decide to add more [domains](https://www.wpbeginner.com/beginners-guide/beginners-guide-what-is-a-domain-name-and-how-do-domains-work/) or IP addresses to the file but ensure to edit the [/etc/hosts](https://www.howtogeek.com/howto/27350/beginner-geek-how-to-edit-your-hosts-file/) file to have those domains point to the local machine (127.0.0.1).
+
+Next will be to generate a key and use the key to generate a [CSR](https://www.thesslstore.com/knowledgebase/ssl-generate/cetificate-signing-request-overview/) (Certificate Signing Request) using the below command.
+
+```bash
+$ openssl genrsa -out localhost.key -des3 2048
+```
+
+The command will generate the localhost private key, and the passphrase will be requested for the key, and the user will be asked to confirm it again.
+
+Next will be to generate CSR using the key, and then the passphrase create above will be requested. Any other details requested can be left as default or keyed in as appropriate. Note the challenge password requested; one can enter anything. Use the below command to generate the CSR:
+
+```bash
+$ openssl req -new -key localhost.key -out localhost.csr
+```
+
+Now with this CSR, we can request the CA to sign a certificate as below:
+
+```bash
+$ openssl x509 -req -in localhost.csr -CA ../cert/CA/CA.pem -CAkey ../cert/CA/CA.key -CAcreateserial -days 3650 -sha256 -extfile localhost.ext -out localhost.crt
+```
+
+This command takes in the CSR (`localhost.csr`), the CA certificate (`CA.pem` and `CA.key`), and the certificate extensions file (`localhost.ext`). Those inputs generate a `localhost.crt` certificate file, valid for ten years.
+
+The server will need the localhost.crt certificate file, and the decrypted key since our localhost.key is in encrypted form. We will need to decrypt the `localhost.key` and store that file too as below:
+
+```bash
+$ openssl rsa -in localhost.key -out localhost.decrypted.key
+```
+
+### Step 3: Creating a Node.js Express Server
+To test the above whether working as expected, we have to create a Node.js Express server that outputs a success message.
+
+Going back to our cert folder, we will initialize a Node.js project and add the express and HTTPS packages using the below commands:
+
+```bash
+$ npm init -y
+$ npm i express https
+$ touch index.js
+```
+
+Then edit the `index.js` file as below:
+
+```JavaScript
+const fs = require('fs');
+const key = fs.readFileSync('../localhost/localhost.decrypted.key');
+const cert = fs.readFileSync('../localhost/localhost.crt');
```

Comment:
Point out that the path for the `localhost.decrypted.key` and `localhost.crt` are dependent on where the reader is running the command from.

For me it was 
```JavaScript
const key = fs.readFileSync('./CA/localhost/localhost.decrypted.key');
const cert = fs.readFileSync('./CA/localhost/localhost.crt');
```